libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2015,104 +2021,173 @@ bool AudioIO::ProcessPlaybackSlices(
-         size_t iChannel = 0;
-         for (; iChannel < nChannels; ++iChannel) {
-            auto &ringBuffer = *mPlaybackBuffers[iBuffer + iChannel];
-            const auto pair = ringBuffer.GetUnflushed(iBlock);
-            // Playback RingBuffers have float format: see AllocateBuffers
-            pointers[iChannel] = reinterpret_cast<float*>(pair.first);
-            // The lengths of corresponding unflushed blocks should be
-            // the same for all channels
-            if (len == 0)
-               len = pair.second;
-            else
-               assert(len == pair.second);
-         }
+      int bufferIndex = 0;
+      for(const auto& seq : mPlaybackSequences)
+      {
+         if(!seq)
+            continue;//no similar check in convert-to-float part
+         const auto channelGroup = seq->FindChannelGroup();
+         if(!channelGroup)
+            continue;
 
          // Are there more output device channels than channels of vt?
          // Such as when a mono sequence is processed for stereo play?
          // Then supply some non-null fake input buffers, because the
          // various ProcessBlock overrides of effects may crash without it.
          // But it would be good to find the fixes to make this unnecessary.
-         float **scratch = &mScratchPointers[mNumPlaybackChannels + 1];
-         while (iChannel < mNumPlaybackChannels)
-            memset((pointers[iChannel++] = *scratch++), 0, len * sizeof(float));
+         auto scratch = &mScratchPointers[mNumPlaybackChannels + 1];
 
-         if (len && pScope) {
-            auto discardable = pScope->Process(*pGroup, &pointers[0],
+         const auto len = mProcessingBuffers[bufferIndex].size();
+         for(unsigned i = 0; i < seq->NChannels(); ++i)
+            pointers[i] = mProcessingBuffers[bufferIndex + i].data();
+         
+         for(unsigned i = seq->NChannels(); i < mNumPlaybackChannels; ++i)
+         {
+            pointers[i] = *scratch++;
+            std::fill_n(pointers[i], mProcessingBuffers[bufferIndex].size(), .0f);
+         }
+
+         if (pScope) {
+            const auto discardable = pScope->Process(channelGroup, &pointers[0],
                mScratchPointers.data(),
                // The single dummy output buffer:
                mScratchPointers[mNumPlaybackChannels],
                mNumPlaybackChannels, len);
-            iChannel = 0;
-            for (; iChannel < nChannels; ++iChannel) {
-               auto &ringBuffer = *mPlaybackBuffers[iBuffer + iChannel];
-               auto discarded = ringBuffer.Unput(discardable);
-               // assert(discarded == discardable);
+            for(int i = 0; i < seq->NChannels(); ++i)
+            {
+               auto& buffer = mProcessingBuffers[bufferIndex + i];
+               buffer.erase(buffer.begin(), buffer.begin() + discardable);
             }
          }
+         bufferIndex += seq->NChannels();
+      }
+   }
+
+   auto samplesAvailable = std::min_element(
```

Comment:
This crashes when playing a MIDI file on an empty project, because `mPlaybackSequences` and, consequently, `mProcessingBuffers`, are then empty.

Reply:
Maybe a solution: the size of `mMasterBuffers` equals the number of playback channels, maybe it could be used instead.
Also, AFAICS these buffers always have the same size, so a simple `mMasterBuffers[0].size()` would be more expressive and sufficient.

Reply:
Or `mPlaybackBuffers`.